k8s/upgrade-job/src/helm/upgrade.rs

```diff
@@ -0,0 +1,197 @@
+        },
+    },
+    helm::{
+        client::{HelmReleaseClient, HelmReleaseClientBuilder},
+        values::generate_values_args,
+    },
+    upgrade,
+};
+use regex::Regex;
+use semver::Version;
+use serde_yaml::Value;
+use snafu::{ensure, ResultExt};
+use std::{fs, path::PathBuf};
+
+/// This is the helm chart variant of the helm chart installed in the cluster.
+/// The PRODUCT may be installed using either of these options, but never both.
+#[derive(Clone, Default)]
+pub(crate) enum HelmChart {
+    #[default]
+    Umbrella,
+    Core,
+}
+
+/// This is a builder for the Helm chart upgrade.
+#[derive(Default)]
+pub(crate) struct HelmUpgradeBuilder {
+    release_name: Option<String>,
+    namespace: Option<String>,
+    umbrella_chart_dir: Option<PathBuf>,
+    core_chart_dir: Option<PathBuf>,
+}
+
+impl HelmUpgradeBuilder {
+    /// This is a builder option to add the Namespace of the helm chart to be upgrade.
+    #[must_use]
+    pub(crate) fn with_namespace<J>(mut self, ns: J) -> Self
+    where
+        J: ToString,
+    {
+        self.namespace = Some(ns.to_string());
+        self
+    }
+
+    /// This is a builder option to add the release name of the helm chart to be upgraded.
+    #[must_use]
+    pub(crate) fn with_release_name<J>(mut self, release_name: J) -> Self
+    where
+        J: ToString,
+    {
+        self.release_name = Some(release_name.to_string());
+        self
+    }
+
+    /// This is a builder option to set the directory path of the Umbrella helm chart CLI option.
+    #[must_use]
+    pub(crate) fn with_umbrella_chart_dir(mut self, dir: Option<PathBuf>) -> Self {
+        self.umbrella_chart_dir = dir;
+        self
+    }
+
+    /// This is a builder option to set the directory path of the Umbrella helm chart CLI option.
+    #[must_use]
+    pub(crate) fn with_core_chart_dir(mut self, dir: Option<PathBuf>) -> Self {
+        self.core_chart_dir = dir;
+        self
+    }
+
+    /// This adds buiilds the HelmUpgrade object.
+    pub(crate) fn build(self) -> Result<HelmUpgrade> {
+        ensure!(
+            self.release_name.is_some() && self.namespace.is_some(),
+            HelmUpgradeOptionsAbsent
+        );
+
+        let release_name = self.release_name.clone().unwrap();
+        // Generate HelmReleaseClient.
+        let client = HelmReleaseClient::builder()
+            .with_namespace(self.namespace.clone().unwrap())
+            .build()?;
+        // Get HelmReleaseElement object for the release specified in CLI options.
+        let chart = client.release_info(release_name.clone())?.chart();
+
+        // Define regular expression to pick out the chart name from the
+        // <chart-name>-<chart-version> string.
+        let umbrella_chart_regex = format!(r"^({UMBRELLA_CHART_NAME}-[0-9]+\.[0-9]+\.[0-9]+)$");
+        // Accepts pre-release and release, both.
+        let core_chart_regex =
+            format!(r"^({CORE_CHART_NAME}-[0-9]+\.[0-9]+\.[0-9]+(-rc\.[0-9]+)?)$");
```

Comment:
this only seems to support rc, I think we should leave it open-ended, we may have beta etc